drivers/photo-electronic/adpd410x/adpd410x.c

```diff
@@ -0,0 +1,660 @@
+			data_byte_count++;
+			break;
+		case 3:
+			data[sample_index] = data_byte_buff[data_byte_count] <<
+					     8;
+			data_byte_count++;
+			data[sample_index] |= data_byte_buff[data_byte_count];
+			data_byte_count++;
+			data[sample_index] |= data_byte_buff[data_byte_count] <<
+					      16;
+			data_byte_count++;
+			break;
+		case 4:
+			data[sample_index] = data_byte_buff[data_byte_count] <<
+					     8;
+			data_byte_count++;
+			data[sample_index] |= data_byte_buff[data_byte_count];
+			data_byte_count++;
+			data[sample_index] |= data_byte_buff[data_byte_count] <<
+					      24;
+			data_byte_count++;
+			data[sample_index] |= data_byte_buff[data_byte_count] <<
+					      16;
+			data_byte_count++;
+			break;
+		}
+		sample_index++;
+		if(((dual_chan & (1 << i)) != 0) && (got_one == 0)) {
+			got_one = 1;
+		} else {
+			i++;
+			got_one = 0;
+		}
+	} while(i < no_slots);
+
+	free(data_byte_buff);
+	free(data_buff);
+	free(slot_bytecount_tab);
+
+	return SUCCESS;
+
+error_data_byte:
+	free(data_byte_buff);
+error_data_buff:
+	free(data_buff);
+error_slot:
+	free(slot_bytecount_tab);
+
+	return FAILURE;
+}
+
+/**
+ * @brief Get a full data packet from the device containing data from all active
+ *        time slots.
+ * @param dev - Device handler.
+ * @param data - Pointer to the data container.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_get_data(struct adpd410x_dev *dev, uint32_t *data)
+{
+	int32_t ret;
+	int8_t i;
+	uint16_t temp_data, dual_chan = 0;
+	uint8_t ts_no;
+
+	ret = adpd410x_reg_read(dev, ADPD410X_REG_OPMODE, &temp_data);
+	if(ret != SUCCESS)
+		return ret;
+	ts_no = ((temp_data & BITM_OPMODE_TIMESLOT_EN) >>
+		 BITP_OPMODE_TIMESLOT_EN) + 1;
+
+	for(i = 0; i < ts_no; i++) {
+		ret = adpd410x_reg_read(dev, (ADPD410X_REG_TS_CTRL_A + i * 0x20),
+					&temp_data);
+		if(ret != SUCCESS)
+			return ret;
+		if((temp_data & BITM_TS_CTRL_A_CH2_EN_A) != 0)
+			dual_chan |= (1 << i);
+	}
+
+	return adpd410x_get_data_packet(dev, data, ts_no, dual_chan);
+}
+
+/**
+ * @brief Setup the device and the driver.
+ * @param device - Pointer to the device handler.
+ * @param init_param - Pointer to the initialization structure.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_setup(struct adpd410x_dev **device,
+		       struct adpd410x_init_param *init_param)
+{
+	int32_t ret;
+	struct adpd410x_dev *dev;
+	uint16_t reg_temp;
+
+	dev = calloc(1, sizeof *dev);
+	if(!dev)
+		return FAILURE;
+
+	dev->dev_type = init_param->dev_type;
+	dev->clk_opt = init_param->clk_opt;
+	dev->ext_lfo_freq = init_param->ext_lfo_freq;
+
+	if(dev->dev_type == ADPD4100)
+		ret = spi_init(&dev->dev_ops.spi_phy_dev,
+			       &init_param->dev_ops_init.spi_phy_init);
+	else
+		ret = i2c_init(&dev->dev_ops.i2c_phy_dev,
+			       &init_param->dev_ops_init.i2c_phy_init);
+	if(ret != SUCCESS)
+		goto error_dev;
+
+	ret = gpio_get(&dev->gpio0, &init_param->gpio0);
```

Comment:
Do these GPIO's have a default state ? If so, you should set it here, before reset.

Reply:
The GPIOs are tristated by default. I included them in case the user would need them.

Reply:
Ok, if they don't need a default state, it's fine.